// react-page-click/lib/PageClick.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _ReactComponentWithPureRenderMixin = require('react/lib/ReactComponentWithPureRenderMixin');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var PageClick = _react2.default.createClass({
  displayName: 'PageClick',

  propTypes: {
    children: _react2.default.PropTypes.node.isRequired,
    onClick: _react2.default.PropTypes.func.isRequired,
    onMouseDown: _react2.default.PropTypes.func,
    onTouchStart: _react2.default.PropTypes.func,
    onMouseUp: _react2.default.PropTypes.func,
    onTouchEnd: _react2.default.PropTypes.func,
    outsideOnly: _react2.default.PropTypes.bool
  },

  getDefaultProps: function getDefaultProps() {
    return {
      outsideOnly: true
    };
  },
  componentWillMount: function componentWillMount() {
    this.insideClick = false;
  },
  componentDidMount: function componentDidMount() {
    global.window.addEventListener('mousedown', this.onDocumentClick, false);
    global.window.addEventListener('touchstart', this.onDocumentClick, false);
  },


  shouldComponentUpdate: _ReactComponentWithPureRenderMixin.shouldComponentUpdate,

  componentWillUnmount: function componentWillUnmount() {
    global.window.removeEventListener('mousedown', this.onDocumentClick, false);
    global.window.removeEventListener('touchstart', this.onDocumentClick, false);
  },
  onDocumentClick: function onDocumentClick() {
    var _props;

    if (this.insideClick) {
      return;
    }
    (_props = this.props).onClick.apply(_props, arguments);
  },
  onMouseDown: function onMouseDown() {
    this.insideClick = true;
    if (this.props.onMouseDown) {
      var _props2;

      (_props2 = this.props).onMouseDown.apply(_props2, arguments);
    }
  },
  onMouseUp: function onMouseUp() {
    this.insideClick = false;
    if (this.props.onMouseUp) {
      var _props3;

      (_props3 = this.props).onMouseUp.apply(_props3, arguments);
    }
  },
  onTouchStart: function onTouchStart() {
    this.insideClick = true;
    if (this.props.onTouchStart) {
      var _props4;

      (_props4 = this.props).onTouchStart.apply(_props4, arguments);
    }
  },
  onTouchEnd: function onTouchEnd() {
    this.insideClick = false;
    if (this.props.onTouchEnd) {
      var _props5;

      (_props5 = this.props).onTouchEnd.apply(_props5, arguments);
    }
  },
  render: function render() {
    var props = this.props.outsideOnly ? {
      onMouseDown: this.onMouseDown,
      onMouseUp: this.onMouseUp,
      onTouchStart: this.onTouchStart,
      onTouchEnd: this.onTouchEnd
    } : {};

    return _react2.default.cloneElement(_react2.default.Children.only(this.props.children), props);
  }
});

exports.default = PageClick;
